### PR TITLE
Issue 326 oxo g fix collapsing

### PR DESCRIPTION
**Pull request by LeeTL1220**
- ``--collapse-number-annotations`` added to CLI  This tells Oncotator to take a known set of numeric columns (listed below) and convert into a single value based on a rule (currently only ``min`` and ``mean`` are supported).  This will eliminate the possibility of a ``|`` appearing in the column as the values will be collapsed according to the rule.

The current columns and rules are found in ``column_collapse.config`` in the source code.  These are key:value of ``annotation_name:method_applied``

Notes:
- There is no easy way for users to modify the list and rules found in ``column_collapse.config``
- Collapsing does not fail if blank or string values are found.  See examples below for behavior.
- "|" is the only delimiter recognized.
- The collapsing rules are driven off of annotation names, not column names in a TCGA MAF.  ``i_`` fields are included only to catch cases where these were specified in the input file.

Collapsing rules as of this writing:
```
# This file is a config file for ColumnCollapser instances.
#
#  key:value is <name of the annotation to collapse>:<method of collapse>
#
#  Currently, the method of collapse can only be: "min" or "mean" (without quotes).  If left blank, mean will be selected.
#
[columns_to_collapse]
ALT_F2R1: min
i_ALT_F2R1: min
REF_F2R1: min
i_REF_F2R1: min
ALT_F1R2: min
i_ALT_F1R2: min
REF_F1R2: min
i_REF_F1R2: min
t_Foxog: mean
i_t_Foxog: mean
i_tumor_f: mean
tumor_f: mean
t_ref_count: min
t_alt_count: min
i_t_ref_count: min
i_t_alt_count: min
n_ref_count: min
n_alt_count: min
i_n_ref_count: min
i_n_alt_count: min
i_tumor_lod_fstar: mean
tumor_lod_fstar: mean
i_t_lod_fstar: mean
t_lod_fstar: mean
gc_content: mean
```

Examples (incl. bad inputs):
mean:
```
0.1|0.2 --> 0.15
4|10 --> 7
4 --> 4
```
min:
```
0.1|0.2 --> 0.1
4|10 --> 4
10 --> 10
```

other examples:
```
# For columns not in the config file
4|6 --> 4|6

# Other examples (regardless of method)
| --> <blank>
|36 --> 36
<blank> --> <blank>
0|blah --> 0|blah
```

- A suffix is specified (``_full``) , which is used to create copies of the original, unaltered annotation.  Note that if the mutation in question does not allow overwriting, a DuplicateAnnotationException will be thrown.

Simple example:
```
# Input
annotation name: t_ref_count
value: 4|10

# Outputs to be rendered in the TCGA MAF 
annotation name: t_ref_count
value: 4

annotation name: t_ref_count_full
value: 4|10
```

### CODE REVIEW

**Comment by LeeTL1220** [comment]
@lbergelson When you get a chance.... thanks

**Comment by LeeTL1220** [comment]
@dheiman Once this pull request is merged, I'll release.

File: oncotator/Oncotator.py [hunk 1]
```diff
@@ -58,7 +58,7 @@
 
 It defines classes_and_methods
 
-@author:     aramos, mgupta, and lichtens
+@author:     aramos, lbergelson, mgupta, and lichtens
```
Comment (lbergelson) [inline oncotator/Oncotator.py]: :+1:

File: oncotator/utils/ColumnCollapser.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        :return: dict with key value pairs of annotation_name:collapsed value.  This can then be used to update the mutation
+        """
+        result = dict()
+        relevant_annotations = self._get_relevant_annotations(mut)
+        for annot in relevant_annotations:
+            result[annot] = self._collapse_column(annot, mut[annot])
+        return result
+
+    def update_mutation(self, mut, new_annotation_source=None, copy_old_suffix=None):
+        """ Given a mutation, update the relevant annotations with new values.  Please note that this
```
Comment (lbergelson) [inline oncotator/utils/ColumnCollapser.py]: Please note that this..... 

Don't leave me hanging here!

File: oncotator/configs/column_collapse.config [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+t_alt_count: min
+i_t_ref_count: min
+i_t_alt_count: min
+n_ref_count: min
+n_alt_count: min
+i_n_ref_count: min
+i_n_alt_count: min
+
+# These columns are for testing
+hamilcar: mean
```
Comment (lbergelson) [inline oncotator/configs/column_collapse.config]: Why are these here?  Shouldn't they go in a test config instead?  If they are here, maybe make them more unlikely somehow?  barca seems close enough to brca that it could somehow sneak into a real file.

File: test/MutationDataTest.py [hunk 4]
```diff
@@ -123,6 +123,13 @@ def testPickleable(self):
         import cPickle
         cPickle.dump(m, open("out/testMDPickle.pkl", 'w'))
 
+    def test_copy(self):
+        """Test annotation copy """
```
Comment (lbergelson) [inline test/MutationDataTest.py]: Could you add some mutation operations to the copy annotation and show that the original is unchanged?  i.e. prove that it's a proper deep copy

File: test/ColumnCollapserTest.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        # m1 = MutationDataFactory.default_create(chr="1", start="10000", end="10000")
+        m1 = MutationData()
+        m1.createAnnotation('ALT_F2R1', "|36", annotationSource="TEST")
+        cc = ColumnCollapser()
+        cc.update_mutation(m1, new_annotation_source="foo", copy_old_suffix="_full")
+        self.assertEqual(m1["ALT_F2R1_full"], "|36")
+        self.assertEqual(m1["ALT_F2R1"], "36")
+        self.assertEqual(m1.getAnnotation("ALT_F2R1_full").getDatasource(), "TEST")
+        self.assertTrue(m1.getAnnotation("ALT_F2R1").getDatasource() != m1.getAnnotation("ALT_F2R1_full").getDatasource())
+
```
Comment (lbergelson) [inline test/ColumnCollapserTest.py]: Could you add a test demonstrating the behavior when the copied annotation conflicts with an existing annotation.  i.e. `depth` is collapsed but `depth_full` already exists.

**Comment by lbergelson** [comment]
@LeeTL1220 Review complete.  :+1:  Just a few extra (hopefully simple) test cases that I think would be good to have.

File: test/ColumnCollapserTest.py (hunk 5 above)
Comment (lbergelson) [inline test/ColumnCollapserTest.py]: I assume this will be resolved according to the output renderer conflict resolution strategy, let me know if you think this is a silly request.

File: oncotator/Oncotator.py (hunk 1 above)
Comment (LeeTL1220) [inline oncotator/Oncotator.py]: Figured it was about time...

File: test/MutationDataTest.py (hunk 4 above)
Comment (LeeTL1220) [inline test/MutationDataTest.py]: This does that already.  ``getAnnotation()`` does not return the value.  It returns an instance of the Annotation class.  Then it is tested for equality.  See ``Annotation.py`` for more info -- it has a custom ``__eq__`` method.

File: test/ColumnCollapserTest.py (hunk 5 above)
Comment (LeeTL1220) [inline test/ColumnCollapserTest.py]: Done with a test that tests for either condition for mutation overwriting.

File: oncotator/utils/ColumnCollapser.py (hunk 2 above)
Comment (LeeTL1220) [inline oncotator/utils/ColumnCollapser.py]: done

File: oncotator/configs/column_collapse.config (hunk 3 above)
Comment (LeeTL1220) [inline oncotator/configs/column_collapse.config]: Removed in oncotator/configs (CLI), but preserved in the test/configs (for automated tests)

**Comment by LeeTL1220** [comment]
@lbergelson Meetings went long today....  Comments addressed.  Please re-review.

@dheiman @ldgauthier @kcibul Looks like I will release tomorrow, not today.  This is the last pull request to be merged before release.

**Comment by lbergelson** [comment]
:+1:
